### PR TITLE
SDF bug fixes

### PR DESCRIPTION
**Pull request by suikki**
I found another stride bug and fixed some docs.

Please check if my fix for issue https://github.com/memononen/SDF/issues/2 makes sense.

I created a font stash fork too and cleaned up my hacks a bit. I'll commit my stuff shortly. It's still not the nicest way to integrate SDF to font stash but at least it's something to play with.

### CODE REVIEW

**Comment by suikki** [comment]
My fonts stash fork for playing around with SDF: https://github.com/suikki/fontstash

I originally hacked it together so that it would always use SDF instead of blurring. I did some cleanup, but there is a lot of room to improve for proper integration. I added a variable that defines if you want blurring or SDF (because I didn't want make too many changes and I didn't want to break the API).
- Added simple support for outlined fonts (without need for a shader) (just because I happened to need it)
- Had to make scratch memory a lot larger (10x) for SDF to work. Not nice if you are not actually using it.
- Added some example/test code for outlining and SDF (not really a good test for SDF drawing because instead of using a shader it's just using alpha testing). In "error" program use left/right to scale rendering and A to toggle alpha testing.

File: src/sdf.h [hunk 1]
```diff
@@ -195,7 +195,13 @@ void sdfBuildDistanceFieldNoAlloc(unsigned char* out, int outstride, float radiu
 	for (y = 1; y < height-1; y++) {
 		for (x = 1; x < width-1; x++) {
 			int tk, k = x + y * stride;
-			if (img[k] > 0 && img[k] < 255) {
+
+			// Fix to https://github.com/memononen/SDF/issues/2
+			if (img[k] == 255) {
```
Comment (memononen) [inline src/sdf.h]: I think we should make this a bit more obvious and have "inline" comment too. i.e.

``` C
// Skip flat areas.
if (img[k] == 0) continue;
if (img[k] == 255) {
    // Special handling for cases where full opaque pixels are next to full transparent pixels.
    // See: https://github.com/memononen/SDF/issues/2
    if ((img[k-1] != 0) && (img[k+1] != 0) && (img[k-stride] != 0) && (img[k+stride] != 0)) continue;
}
...
```

**Comment by memononen** [comment]
Good PR, I think the special case should be made a bit more obvious.

I tested an alternative method too, which surfaced another bug. Spot the difference here vs. the repo :) (hint somewhere in gy):

``` C
gx = -(float)img[k-stride-1] - SDF_SQRT2*(float)img[k-1] - (float)img[k+stride-1] + (float)img[k-stride+1] + SDF_SQRT2*(float)img[k+1] + (float)img[k+stride+1];
gy = -(float)img[k-stride-1] - SDF_SQRT2*(float)img[k-stride] - (float)img[k-stride+1] + (float)img[k+stride-1] + SDF_SQRT2*(float)img[k+stride] + (float)img[k+stride+1];
```

With the above fix in place, it is possible to change the skip logic to this:

``` C
...
d = sdf__edgedf(gx, gy, (float)img[k]/255.0f);
// Skip if the projected point on iso-line is way outside the pixel. 
if (d > 0.50f || d < -0.50f) continue;
...
```

In addition we could add early out skip for `img[k] == 0` before the gradient calculations.

**Comment by suikki** [comment]
Yeah, nice find.

Could there still be some other issues in the algorithm? I tried with your proposed fix and even though the resulting rendered font looks quite good, the distance field itself has some suspicious looking artifacts:

artifacts

**Comment by memononen** [comment]
That includes both the `gy` bug fix and the `d` based skip?

**Comment by suikki** [comment]
Yes, both. And I removed my earlier fix.

**Comment by memononen** [comment]
How about if you add `if (img[k] == 0) continue;` before the gx,gy calculations.

**Comment by suikki** [comment]
This was the code I used:

```
        if (img[k] > 0) {
            struct SDFpoint c = { (float)x, (float)y };
            float d, gx, gy, glen;
            // Calculate gradient direction

            gx = -(float)img[k-stride-1] - SDF_SQRT2*(float)img[k-1] - (float)img[k+stride-1] + (float)img[k-stride+1] + SDF_SQRT2*(float)img[k+1] + (float)img[k+stride+1];
            gy = -(float)img[k-stride-1] - SDF_SQRT2*(float)img[k-stride] - (float)img[k-stride+1] + (float)img[k+stride-1] + SDF_SQRT2*(float)img[k+stride] + (float)img[k+stride+1];

            if (fabsf(gx) < 0.001f && fabsf(gy) < 0.001f) continue;
            glen = gx*gx + gy*gy;
            if (glen > 0.0001f) {
                glen = 1.0f / sqrtf(glen);
                gx *= glen;
                gy *= glen;
            }
            // Find nearest point on contour.
            tk = x + y * width;
            d = sdf__edgedf(gx, gy, (float)img[k]/255.0f);
            if (d > 0.50f || d < -0.50f) continue;

            tpt[tk].x = x + gx*d;
            tpt[tk].y = y + gy*d;
            tdist[tk] = sdf__distsqr(&c, &tpt[tk]);
        }
```

So I had the earlier _if (img[k] > 0)_ there. I just tried removing it but I think the artifacts are still there. Looks a bit different though

**Comment by suikki** [comment]
The screencap is from modified font stash error.c test program that you can run too, if you want:
https://github.com/suikki/fontstash

**Comment by memononen** [comment]
ok, that change looks as it should. How does that same distance field look with your proposed fix?

**Comment by suikki** [comment]
Some smaller artifacts in the bottom?

artifacts3

**Comment by suikki** [comment]
And that was my earlier fix + gy fix

**Comment by suikki** [comment]
I added a couple of features to the SDF example program for easier testing: https://github.com/suikki/SDF/commit/b71d84bc7f4f34d3d6532e1a8e25211f60b7c5c4

The included test image contains text and some lines with hard edges. The image also has the same line patterns shifted by half a pixel on both horizontally and vertically (antialiased).

**Comment by memononen** [comment]
I had little time to take a look at this. Can you try the following:

``` C
            // Skip flat areas.
            if (img[k] == 255) continue;
            if (img[k] == 0) {
                // Special handling for cases where full opaque pixels are next to full transparent pixels.
                // See: https://github.com/memononen/SDF/issues/2
                int he = img[k-1] == 255 || img[k+1] == 255;
                int ve = img[k-stride] == 255 || img[k+stride] == 255;
                if (!he && !ve) continue;
            }
```

The skip logic changed a bit. There was something odd about your skip logic, I cannot quite figure out what it was, but the above seems to work better. Secondly, doing the logic on zero pixels (that is, outside), also allows to handle one pixel features.

**Comment by suikki** [comment]
Yeah, definitely seems to work best out of the fixes so far.

There are still small artifacts than that seem to appear on vertical edges near a corner. Could be some unrelated issue.

artifacts4

**Comment by memononen** [comment]
How about if you include diagonals in the check too?

``` C
                int he = img[k-1] == 255 || img[k+1] == 255;
                int ve = img[k-stride] == 255 || img[k+stride] == 255;
                int d1e = img[k-stride-1] == 255 || img[k+stride+1] == 255;
                int d2e = img[k+stride-1] == 255 || img[k-stride+1] == 255;
                if (!he && !ve && !d1e && !d2e) continue;
```

**Comment by suikki** [comment]
It didn't help.

I decided to dig a bit deeper and actually try to understand what's happening. I can't believe I found the bug :D

It was indeed a separate issue. There is an extra `-1` on this line:
https://github.com/memononen/SDF/blob/master/src/sdf.h#L242

After fixing that I tried including the diagonals again. It changes the output just slightly, but I can't really say which looks better.

**Comment by memononen** [comment]
Excellent find! Let's get this merged, and I can fix the other function too. Thanks for looking into this!

**Comment by suikki** [comment]
Great, thanks.

I think you can close the issues too.

**Pull request merged**